packages/credential-provider-http/src/fromHttp/retry-wrapper.ts

```diff
@@ -10,14 +12,19 @@ export interface RetryableProvider<T> {
  */
 export const retryWrapper = <T>(
   toRetry: RetryableProvider<T>,
-  maxRetries: number,
-  delayMs: number
+  options: { maxRetries: number; delayMs: number; logger?: Logger }
 ): RetryableProvider<T> => {
+  const { maxRetries, delayMs, logger } = options;
   return async () => {
     for (let i = 0; i < maxRetries; ++i) {
       try {
         return await toRetry();
       } catch (e) {
+        if (i === maxRetries - 1) {
+          logger?.warn?.(
+            `@aws-sdk/credential-provider-http - retry limit ${maxRetries} reached. error: ${JSON.stringify(e)}`
+          );
```

Comment:
This is not the right place to emit a warning because there is one more attempt made on line 31. For maxRetries=2, this would be attempt 2 of 3 total.

Also, this wouldn't be a scenario for which to emit a warning. The default credential chain can fail at many steps and be intentional. 

To ensure that a specific credential source throws a highly visible Error, create a chain or single-step function in which it is the final credential source.

```js
// example: use critical credential source as final step in a credential chain.
import { fromEnv, fromHttp } from "@aws-sdk/credential-providers";

new SDKClient({
  credentials: async () => fromEnv({})().catch(fromHttp({}))
});
```

If you want to continue using the default provider chain, use a log filter to surface the `CredentialProviderError`s:

```js
// example: log CredentialProviderError indicating progression in the 
//                 default credential chain from one source to the next.
new SDKClient({
  logger: {
     ...console,
    debug(...args) {
      // surface debug messages within the credential chain coming from CredentialProviderError
      if (args[0].startsWith("@smithy/property-provider")) {
        console.warn(...args);
      }
    },
  }
});
```